src/Processors/Transforms/TotalsHavingTransform.cpp

```diff
@@ -17,33 +17,39 @@ namespace ErrorCodes
     extern const int ILLEGAL_COLUMN;
 }
 
-void finalizeChunk(Chunk & chunk)
+void finalizeChunk(Chunk & chunk, const ColumnNumbers & aggregation_keys)
 {
     auto num_rows = chunk.getNumRows();
     auto columns = chunk.detachColumns();
 
-    for (auto & column : columns)
-        if (typeid_cast<const ColumnAggregateFunction *>(column.get()))
-            column = ColumnAggregateFunction::convertToValues(IColumn::mutate(std::move(column)));
+    for (size_t i = 0; i < columns.size(); ++i)
+    {
+        if (std::find(aggregation_keys.begin(), aggregation_keys.end(), i) != aggregation_keys.end())
+            continue;
+
+        auto & column = columns[i];
+        column = ColumnAggregateFunction::convertToValues(IColumn::mutate(std::move(column)));
+    }
 
     chunk.setColumns(std::move(columns), num_rows);
 }
 
-void finalizeBlock(Block & block)
+static void finalizeBlock(Block & block, const ColumnNumbers & aggregation_keys)
 {
     for (size_t i = 0; i < block.columns(); ++i)
     {
+        if (std::find(aggregation_keys.begin(), aggregation_keys.end(), i) != aggregation_keys.end())
+            continue;
+
         ColumnWithTypeAndName & current = block.getByPosition(i);
         const DataTypeAggregateFunction * unfinalized_type = typeid_cast<const DataTypeAggregateFunction *>(current.type.get());
+        assert(unfinalized_type);
```

Comment:
maybe cast to a reference and get an exception if smth went wrong?

Reply:
Good suggestion, done, thanks.